pkg/ccl/changefeedccl/changefeed_processors.go

```diff
@@ -1941,16 +1948,214 @@ func (cf *changeFrontier) manageProtectedTimestamps(
+		newPtsTimestamp, updatedPerTablePts, err := cf.managePerTableProtectedTimestamps(ctx, txn, progress, &ptsEntries)
+		if err != nil {
+			return false, err
+		}
+		return cf.advanceProtectedTimestampToTimestamp(ctx, txn, progress, pts, newPtsTimestamp, updatedPerTablePts)
+	}
+
+	if len(ptsEntries.ProtectedTimestampRecords) > 0 {
+		// If we're not using per table protected timestamps to protect lagging tables,
+		// we will rely on the single protected timestamp record to protect all tables,
+		// but we need to release all the per table PTS records that we created earlier.
+		if err := cf.releasePerTableProtectedTimestampRecords(ctx, txn, &ptsEntries); err != nil {
+			return false, err
+		}
+	}
+
+	highwater := cf.frontier.Frontier()
+	if highwater.Less(cf.highWaterAtStart) {
+		highwater = cf.highWaterAtStart
+	}
+	return cf.advanceProtectedTimestampToTimestamp(ctx, txn, progress, pts, highwater, false /* updatedPerTablePts */)
+}
+
+func (cf *changeFrontier) managePerTableProtectedTimestamps(
+	ctx context.Context,
+	txn isql.Txn,
+	progress *jobspb.ChangefeedProgress,
+	ptsEntries *changefeedpb.ProtectedTimestampRecords,
+) (newPtsTimestamp hlc.Timestamp, updatedPerTablePts bool, err error) {
+	// TODO: This seems weird, come back to it. Is there a better way to do this?
+	var leastLaggingTimestamp hlc.Timestamp
+	for _, frontier := range cf.frontier.Frontiers() {
+		if leastLaggingTimestamp.IsEmpty() || frontier.Frontier().After(leastLaggingTimestamp) {
+			leastLaggingTimestamp = frontier.Frontier()
+		}
+	}
+
+	// We use this to know if we need to update the per table PTS records
+	// in the job info protobuf.
+	perTablePTSChanged := false
+	newPtsTimestamp = leastLaggingTimestamp
+
 	pts := cf.FlowCtx.Cfg.ProtectedTimestampProvider.WithTxn(txn)
+	for tableId, frontier := range cf.frontier.Frontiers() {
+		tableHighWater := frontier.Frontier()
+		if tableHighWater.IsEmpty() {
+			// TODO: What should I do for this case? Will this come up outside of tests?
+			// For example, what if this table was recently added, is this the place
+			// to create the per-table PTS record for it?
+			fmt.Println("tableHighWater is empty for tableId", tableId)
+			continue
+		}
+		lagDuration := changefeedbase.ProtectTimestampBucketingInterval.Get(&cf.FlowCtx.Cfg.Settings.SV)
+		isLagging := !tableHighWater.AddDuration(lagDuration).After(leastLaggingTimestamp)
+
+		if cf.knobs.IsTableLagging != nil && cf.knobs.IsTableLagging(tableId) {
+			isLagging = true
+		}
+
+		if isLagging {
+			if ptsEntries.ProtectedTimestampRecords[tableId] != nil {
+				if updated, err := cf.advancePerTableProtectedTimestampRecord(ctx, txn, progress, ptsEntries, tableId, tableHighWater, pts); err != nil {
+					return hlc.Timestamp{}, false, err
+				} else if updated {
+					updatedPerTablePts = true
+				}
+			} else {
+				if err := cf.createPerTableProtectedTimestampRecord(ctx, txn, progress, ptsEntries, tableId, tableHighWater, pts); err != nil {
```

Comment:
Also discussed offline but leaving a note here that I continue to feel strongly that we should not include all the system tables in every per-table PTS record that we create but rather have one system-table-only PTS record like I commented in the design doc. If you decide not to do that in this PR, could you file an issue for it and leave TODO comments with the issue number in the code?

Reply:
Totally fine to leave it as a seperate todo, so long as no production changefeed will ever enter this immediate state, where there are O(numTable) pts's on the system tables.